network/src/peer_store/sqlite/peer_store.rs

```diff
@@ -290,36 +303,72 @@ impl PeerStore for SqlitePeerStore {
         }
         peers
     }
-    fn peer_addrs<'a>(&'a self, peer_id: &'a PeerId, count: u32) -> Option<Vec<Multiaddr>> {
-        self.get_peer_info(peer_id).map(|peer| {
-            db::PeerAddr::get_addrs(&self.conn, peer.id, count).expect("get peer addrs")
-        })
+
+    fn peer_addrs<'a>(&'a self, peer_id: &'a PeerId, count: u32) -> Vec<PeerAddr> {
+        db::PeerAddrDB::get_addrs(&self.conn, peer_id, count).expect("get peer addrs")
     }
 
-    fn peers_to_attempt(&self, count: u32) -> Vec<(PeerId, Multiaddr)> {
+    fn peers_to_attempt(&self, count: u32) -> Vec<PeerAddr> {
+        let now_ms = unix_time_as_millis();
         let peers = db::get_peers_to_attempt(&self.conn, count).expect("get peers to attempt");
-        self.find_addrs_for_peers(&self.conn, peers)
-            .expect("find_addrs_for_peers failed")
+        peers
+            .into_iter()
+            .filter_map(|peer_id| {
+                let mut paddrs = db::PeerAddrDB::get_addrs(&self.conn, &peer_id, DEFAULT_ADDRS)
+                    .expect("get peer addr");
+                let mut rng = thread_rng();
+                // randomly find a address to attempt
+                paddrs.shuffle(&mut rng);
+                paddrs
+                    .into_iter()
+                    .find(|paddr| !self.is_addr_banned(&paddr.addr) && !paddr.is_terrible(now_ms))
+            })
+            .collect()
     }
 
-    fn peers_to_feeler(&self, count: u32) -> Vec<(PeerId, Multiaddr)> {
-        let peers = db::get_peers_to_feeler(
-            &self.conn,
-            count,
-            unix_time() - Duration::from_secs(LAST_CONNECTED_TIMEOUT_SECS),
-        )
-        .expect("get peers to feeler");
-        self.find_addrs_for_peers(&self.conn, peers)
-            .expect("find_addrs_for_peers failed")
+    fn peers_to_feeler(&self, count: u32) -> Vec<PeerAddr> {
+        let now_ms = unix_time_as_millis();
+        let peers = db::get_peers_to_feeler(&self.conn, count, now_ms - ADDR_TIMEOUT_MS)
+            .expect("get peers to feeler");
+        peers
+            .into_iter()
+            .filter_map(|peer_id| {
+                let mut paddrs = db::PeerAddrDB::get_addrs(&self.conn, &peer_id, DEFAULT_ADDRS)
+                    .expect("get peer addr");
+                // find worst addr to feeler unless it is terrible or banned or already tried in one minute
+                paddrs.sort_by(|a, b| a.last_connected_at_ms.cmp(&b.last_connected_at_ms));
+                paddrs.into_iter().find(|paddr| {
+                    if paddr.is_terrible(now_ms) {
+                        // delete terrible addr
+                        db::PeerAddrDB::delete(&self.conn, &paddr.peer_id, &paddr.addr)
+                            .expect("delete peer addr");
+                        false
+                    } else {
+                        !self.is_addr_banned(&paddr.addr)
+                            && paddr.last_tried_at_ms < now_ms - 60_000
```

Comment:
refactor this to `PeerAddr` fn, and reuse it in `is_terrible` fn, which can remove the duplicate code and two hardcoded value.